frontend/src/components/NewRunParameters.tsx

```diff
@@ -39,23 +65,143 @@ class NewRunParameters extends React.Component<NewRunParametersProps> {
-            {initialParams.map((param, i) => (
-              <TextField
-                id={`newRunPipelineParam${i}`}
-                key={i}
-                variant='outlined'
-                label={param.name}
-                value={param.value || ''}
-                onChange={ev => handleParamChange(i, ev.target.value || '')}
-                style={{ maxWidth: 600 }}
-                className={commonCss.textField}
-              />
-            ))}
+            {initialParams.map((param, i) => {
+              return (
+                <ParamEditor
+                  key={i}
+                  id={`newRunPipelineParam${i}`}
+                  onChange={(value: string) => handleParamChange(i, value)}
+                  param={param}
+                />
+              );
+            })}
           </div>
         )}
       </div>
     );
   }
 }
 
+interface ParamEditorProps {
+  id: string;
+  onChange: (value: string) => void;
+  param: ApiParameter;
+}
+
+interface ParamEditorState {
+  isEditorOpen: boolean;
+  isInJsonForm: boolean;
+  isJsonField: boolean;
+}
+
+class ParamEditor extends React.Component<ParamEditorProps, ParamEditorState> {
+  public static getDerivedStateFromProps(
+    nextProps: ParamEditorProps,
+    prevState: ParamEditorState,
+  ): { isInJsonForm: boolean; isJsonField: boolean } {
+    let isJson = true;
+    try {
+      const displayValue = JSON.parse(nextProps.param.value || '');
+      // Nulls, booleans, strings, and numbers can all be parsed as JSON, but we don't care
+      // about rendering. Note that `typeOf null` returns 'object'
+      if (displayValue === null || typeof displayValue !== 'object') {
+        throw new Error('Parsed JSON was neither an array nor an object. Using default renderer');
+      }
+    } catch (err) {
+      isJson = false;
+    }
+    return {
+      isInJsonForm: isJson,
+      isJsonField: prevState.isJsonField || isJson,
+    };
+  }
+
+  constructor(props: any) {
+    super(props);
+
+    this.state = {
+      isEditorOpen: false,
+      isInJsonForm: false,
+      isJsonField: false,
+    };
+  }
+
+  public render(): JSX.Element | null {
+    const { id, onChange, param } = this.props;
+
+    const onClick = () => {
+      if (this.state.isInJsonForm) {
+        const displayValue = JSON.parse(param.value || '');
+        if (this.state.isEditorOpen) {
+          onChange(JSON.stringify(displayValue) || '');
+        } else {
+          onChange(JSON.stringify(displayValue, null, 2) || '');
+        }
+      }
+      this.setState({
+        isEditorOpen: !this.state.isEditorOpen,
+      });
+    };
+
+    return (
+      <>
+        {this.state.isJsonField ? (
+          <TextField
+            id={id}
+            disabled={this.state.isEditorOpen}
+            variant='outlined'
+            label={param.name}
+            value={param.value || ''}
+            onChange={ev => onChange(ev.target.value || '')}
+            style={{ maxWidth: 600 }}
```

Comment:
nit: Please write this as a css
you can follow the pattern https://github.com/kubeflow/pipelines/blob/e824c0b2f63a8891a18f437a1ad2b94990dd67a9/frontend/src/components/SideNav.tsx#L262 to concatenate two class names